ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -26,297 +17,401 @@ const SHORT_INSTR_MAX_OPERANDS: u64 = 3;
 ///  - `*_SHIFT`: the number of bits required to left shift a field's value into position (from the
 ///  LSB).
 ///
-/// Bit fiddling for a short operands:
-const SHORT_OPERAND_SIZE: u64 = 18;
-const SHORT_OPERAND_KIND_SIZE: u64 = 3;
-const SHORT_OPERAND_KIND_MASK: u64 = 7;
-const SHORT_OPERAND_VALUE_SIZE: u64 = 15;
-const SHORT_OPERAND_VALUE_SHIFT: u64 = SHORT_OPERAND_KIND_SIZE;
-const SHORT_OPERAND_MAX_VALUE: u64 = !(u64::MAX << SHORT_OPERAND_VALUE_SIZE);
-const SHORT_OPERAND_MASK: u64 = 0x3ffff;
-/// Bit fiddling for instructions.
-const INSTR_ISSHORT_SIZE: u64 = 1;
-const INSTR_ISSHORT_MASK: u64 = 1;
-/// Bit fiddling for short instructions.
-const SHORT_INSTR_OPCODE_MASK: u64 = 0xe;
-const SHORT_INSTR_OPCODE_SHIFT: u64 = INSTR_ISSHORT_SIZE;
-const SHORT_INSTR_FIRST_OPERAND_SHIFT: u64 = INSTR_ISSHORT_SIZE + OPCODE_SIZE;
-
-/// An instruction is identified by its index in the instruction vector.
-#[derive(Copy, Clone)]
-pub(crate) struct InstructionID(usize);
-
-impl InstructionID {
-    pub(crate) fn new(v: usize) -> Self {
-        Self(v)
-    }
+const OPERAND_INDEX_MASK: u16 = 0x7fff;
+
+// The maximum size of an operand we can encode in an instruction is 15 bits.
+const MAX_OPERAND_IDX: u16 = 32768;
+
+// Pre-allocated type indices.
+const TYPE_INDEX_VOID: u16 = 0;
+const TYPE_INDEX_PTR: u16 = 1;
+const TYPE_INDEX_UNIMPLEMENTED: u16 = 2;
+
+/// A type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum Type {
+    Void,
+    Integer(IntegerType),
+    Ptr,
+    Func(FuncType),
+    Struct(StructType),
+    Unimplemented,
+}
 
-    pub(crate) fn get(&self) -> usize {
-        self.0
-    }
+/// An integer type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct IntegerType {
+    // FIXME: todo.
 }
 
-/// An operand kind.
-#[repr(u64)]
-#[derive(Debug, FromRepr, PartialEq)]
-pub enum OpKind {
-    /// The operand is not present.
-    ///
-    /// This is used in short instructions where 3 operands are inlined. If the instruction
-    /// requires fewer then 3 operands, then it can use this variant to express that.
-    ///
-    /// By using the zero discriminant, this means that a freshly created short instruction has
-    /// with zero operands until they are explicitly filled in.
-    NotPresent = 0,
-    /// The operand references a previously defined local variable.
-    Local,
+/// A structure type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct StructType {
+    // FIXME: todo.
 }
 
-impl From<u64> for OpKind {
-    fn from(v: u64) -> Self {
-        // unwrap safe assuming only valid discriminant numbers are used.
-        Self::from_repr(v).unwrap()
-    }
+/// A function type.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct FuncType {
+    /// Type indices for the function's formal arguments.
+    arg_tys: Vec<TypeIndex>,
```

Comment:
I'm fine with using `ty` as the abbrevation, but then I wondered if we should use the same abbreviation in the struct names too? [I don't feel strongly about this, so feel free to ignore it!]

Reply:
This is something we are quite guilty of across our code base. We don't have very unified terminology.

Examples:
 - `idx` vs `index`
 -  `func` vs `function`
 - `const` vs `constant`
 
It bothers me.

Note that `type` is special, in that it's a keyword in the rust grammar.

I guess we should use `ty` consistently.

Reply:
There's the repo as a whole, but let's at least make this part of the codebase (at least the bits added by this PR) consistent.

[Later I realised that `ty` and `tyidx` should probably be different things.]

Reply:
> [Later I realised that `ty` and `tyidx` should probably be different things.]

Yes, so in this case it should be `arg_tyidx`, or just `tyidx`. I think it's important to differentiate between an actual `Type` or its `TypeIndex`.